### DISCUSSION TITLE
Github Actions is buffering output and not displaying step progress in real time

### DISCUSSION THREAD

**Discussion by iMerica** [community/community#52284] [Actions]
### Select Topic Area

Bug

### Body

Hi, 

I've confirmed [this issue](https://stackoverflow.com/questions/64367788/avoid-github-action-buffering) is happening in self-hosted runners and Github's Cloud runners.  This is a shame because even really outdated CI platforms like Bitbucket and Jenkins display build step output in real-time. 

Is there anything on the road map to view step progress in real time?

Thank you,

Michael

**Comment by IanButterworth** [comment]
I am surprised this issue remains. It's very frustrating when trying to review long running jobs. The only workaround I am aware of is to open a browser window for the run at the start, but it's awkward.

**Comment by aleric-infraspec** [comment]
I still have this issue in 2024, with the github hosted runners.

**Comment by AlyaGomaa** [comment]
Happening to me with github cloud runners! any updates?

**Comment by bholzer** [comment]
Still a problem for me. How else am I supposed to monitor the progress of a job?

**Comment by KotM** [comment]
![image](https://github.com/user-attachments/assets/daf2b1ce-705c-4ccd-a8c8-f8c92b815e31)

Is it what you're looking for? Just add `environment: Test` below `runs-on` and see the progress bar appearing. It is curious, but it is working. :)

**Reply by bholzer** [reply]
It's not the progress bar that I'm missing; that works for me.

My script that runs in a step writes output to STDOUT, but that output does not appear until the step is fully complete.

**Reply by KotM** [reply]
Interesting... What about the output from other scripts?

For example, I don't observe such behavior (with the "windows-latest" runner, if that makes sense) for any of my steps. Installing MySQL, building/testing a project -- everything produces output line by line for each step inside.

**Comment by diofeher** [comment]
I'm seeing the same here on ubuntu-latest on Github Cloud runners

**Comment by ssbarnea** [comment]
That is a very old issue with github runners and the behaviour can be described as

1. OFTEN stderr and stdout lines are out of sync due to buffering which does not flush each line, making hard to debug failures as lines order on output is not correct
2. OCCASIONAL lines are missing from output, missing last error when a program exists

That applies to default shells (usually bash), but for python programs there is a workaround, define `PYTHONUNBUFFERED=1`.

**Comment by miberecz** [comment]
So far, any other platform I used had this function. Gitlab, Jenkins etc... I do not see why is it an issue here

**Comment by Kuberwastaken** [comment]
Haha wow, this has been an issue for long, facing the same now

**Comment by Eidan-Eon** [comment]
What is the ETA for a fix?
